Translate the new representation of Intel FPGA IVDep loop attribute

AGindinson opened the pull request:
The new LLVM IR representation of [[intelfpga::ivdep]] takes into account
the array variables being accessed from within the loop. A brief outline is
that each array variable's GetElementPtr instruction is marked with a
so-called "index group metadata". These index groups are referenced by
IVDep-specific "llvm.loop.parallel_access_indices", which is (finally!)
attached to "llvm.loop" metadata.

The SPIR-V specification for this version of the attribute can be found at:
https://github.com/KhronosGroup/SPIRV-Registry/blob/master/extensions/INTEL/SPV_INTEL_fpga_loop_controls.asciidoc

Some outline of refactoring that took place:
1) The getLoopControl() utility was moved from SPIRVUtil into SPIRVWriter,
the helper being Writer-specific anyway.
2) Several NFCs regarding getMDOperandAs...() helpers now allow safe and
convenient extraction of integer values and MDNode objects from the
"parent" metadata node.
3) Upon encountering loop control parameters, SPIRVReader is now
required to perform full loop analysis for the given function. This
is required to allow convenient access to the GEP instructions in
all of the embedded loops' body blocks.

The older representation of IVDep via {!llvm.loop.ivdep.*} metadata
was retained to ensure backwards compatibility. The unit testing that
used to rely on this representation while not being directly related
aimed at testing the IVDep was reworked to use the other loop attributes.

Signed-off-by: Artem Gindinson <artem.gindinson@intel.com>

AGindinson commented:
Tagging:
@AlexeySotkin @AlexeySachkov @MrSidims @mlychkov @vmaksimo @erichkeane

MrSidims commented on test/transcoding/FPGAIVDepLoopAttr.ll:
```diff
@@ -0,0 +1,416 @@
+11:                                               ; preds = %29, %0
+  %12 = load i32, i32* %5, align 4, !tbaa !9
+  %13 = icmp ne i32 %12, 10
+  ; CHECK-SPIRV: LoopMerge [[CONTINUE_BLOCK:[0-9]+]] {{[0-9]+}} {{[0-9]+}} 5891 4 [[ARRAY_A]] 5 [[ARRAY_D]] 5 [[ARRAY_B]] 6 [[ARRAY_C]] 0
```
Something wrong with LoopMerge. It shall be:
`LoopMerge {{Merge Block}} {{Continue Target}} [[Loop Control = 5891]] ...`
but here I see three numbers before Loop Control, or I'm missing something?

AGindinson replied:
It's more so that I've obfuscated the literal of the ExtendedLoopControl under the third number. These lines need to be reworked into:
```suggestion
  ; CHECK-SPIRV: LoopMerge [[MERGE_BLOCK:[0-9]+]] {{[0-9]+}} 2147483648 5891 4 [[ARRAY_A]] 5 [[ARRAY_D]] 5 [[ARRAY_B]] 6 [[ARRAY_C]] 0
```

MrSidims commented on lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -685,49 +689,145 @@ void SPIRVToLLVM::setLLVMLoopMetadata(const LoopInstType *LM, Instruction *BI) {
+        // A single run over the loop to retrieve all GetElementPtr instructions
+        // that access relevant array variables
+        std::map<Value *, std::vector<GetElementPtrInst *>> ArrayGEPMap;
+        for (auto &BB : LoopObj->blocks())
```
Please, put a bracket here.

MrSidims commented on lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -685,49 +689,145 @@ void SPIRVToLLVM::setLLVMLoopMetadata(const LoopInstType *LM, Instruction *BI) {
+        std::map<Value *, std::vector<GetElementPtrInst *>> ArrayGEPMap;
+        for (auto &BB : LoopObj->blocks())
+          for (Instruction &I : *BB) {
+            auto *GEP = dyn_cast<GetElementPtrInst>(&I);
```
```suggestion
            if (auto *GEP = dyn_cast<GetElementPtrInst>(&I)) {
```

AGindinson replied:
The variable is re-used below, so this isn't going to work out if we're looking to retain the early `continue`.

MrSidims commented on lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -685,49 +689,145 @@ void SPIRVToLLVM::setLLVMLoopMetadata(const LoopInstType *LM, Instruction *BI) {
 }
 
-void SPIRVToLLVM::transLLVMLoopMetadata(const Function *F) {
+void SPIRVToLLVM::transLLVMLoopMetadata(Function *F) {
```
That is looking fishy, why is it no longer const?

AGindinson replied:
The DominatorTree analysis does not accept a const object for reasons I was unable to determine. That said, would a `const_cast` in the `DominatorTree` object initialization make things a little better?

AGindinson replied:
I've decided to go with the `const_cast`. It used to be the choice for an analogous situation at the previous version of line  719

MrSidims commented on lib/SPIRV/SPIRVUtil.cpp:
```diff
@@ -848,21 +848,28 @@ ConstantInt *getSizet(Module *M, uint64_t Value) {
 ///////////////////////////////////////////////////////////////////////////////
 int getMDOperandAsInt(MDNode *N, unsigned I) {
-  return mdconst::dyn_extract<ConstantInt>(N->getOperand(I))->getZExtValue();
+  auto *Value = mdconst::dyn_extract_or_null<ConstantInt>(N->getOperand(I));
```
```suggestion
  if (auto *Value = mdconst::dyn_extract_or_null<ConstantInt>(N->getOperand(I)))
    return Value->getZExtValue();
  return 0;
```

AGindinson replied:
Do I understand correctly that this aims at making the code more explicit? (i.e. I'm totally down with this reasoning)
What concerns me most about this particular function is whether it's a good idea to return 0 in case the `MDOperand` is not an integer value, I'll see if there's a way to avoid this and still fulfil my purpose in `LLVMLoopParallelAccessIndices::initialize()`

AGindinson replied:
I've decided to restore the original version of the function and do the required "custom" extraction of the safelen value directly.

MrSidims commented on lib/SPIRV/SPIRVUtil.cpp:
```diff
@@ -848,21 +848,28 @@ ConstantInt *getSizet(Module *M, uint64_t Value) {
-  Metadata *Op = N->getOperand(I);
-  if (!Op)
+std::string getMDOperandAsString(MDNode *N, unsigned I) {
+  MDString *Str = dyn_cast_or_null<MDString>(getMDOperandOrNull(N, I));
```
```suggestion
  if (auto *Str = dyn_cast_or_null<MDString>(getMDOperandOrNull(N, I)))
    return Str->getString().str();
  return "";
```

MrSidims commented on lib/SPIRV/SPIRVWriter.cpp:
```diff
@@ -705,6 +705,169 @@ SPIRV::SPIRVInstruction *LLVMToSPIRV::transUnaryInst(UnaryInstruction *U,
 
+/// Check if the module contains llvm.loop.* metadata
+static bool hasLoopMetadata(const Module *M) {
+  for (const Function &F : *M)
```
I would love to see brackets in such cases, but it's up to you to apply this comment (as a comment from above about brackets).

AGindinson replied:
When starting out on my LLVM journey, I couldn't accept the braces' absence myself. But then came a review request to remove these... I'll gladly apply this suggestion 👍

AGindinson commented on lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -619,8 +619,12 @@ void SPIRVToLLVM::setLLVMLoopMetadata(const LoopInstType *LM, Instruction *BI) {
     Metadata.push_back(getMetadataFromName("llvm.loop.unroll.enable"));
   else if (LC & LoopControlDontUnrollMask)
     Metadata.push_back(getMetadataFromName("llvm.loop.unroll.disable"));
+  // TODO: The token has been deprecated in LLVM IR. Should be deprecated
```
After some thought, I probably wouldn't want to leave `TODO`s that question the spec.

mlychkov commented on lib/SPIRV/SPIRVInternal.h:
```diff
@@ -774,6 +774,9 @@ int getMDOperandAsInt(MDNode *N, unsigned I);
 std::string getMDOperandAsString(MDNode *N, unsigned I);
 
+/// Get metadata operand as another metadata node
+MDNode *getMDOperandAsMDNode(MDNode *N, unsigned I);
```
It would be nice to move common improvements into additional commit to separate them from ivdep feature functionality. But it is optional.

AGindinson replied:
I can do this to simplify the reviewers' life. Would this include the transition of LoopControl-specific functions into `SPIRVWriter`?

mlychkov replied:
As all commits are squashed before to be merged then no more sense in this change. Drop it.

AlexeySotkin replied:
Not all commits are squashed. In this repo we are trying to maintain linear history. So normally we push `Rebase and Merge` button which adds all commits from the PR on top of the target branch(master).

mlychkov commented on lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -685,49 +689,145 @@ void SPIRVToLLVM::setLLVMLoopMetadata(const LoopInstType *LM, Instruction *BI) {
                           LoopControlParameters[++NumParam])));
         break;
       }
-      default:
```
>   if (LC & LoopControlPeelCountMask) {
    ++NumParam;
    assert(NumParam <= LoopControlParameters.size() &&
           "Missing loop control parameter!");
  }
  if (LC & LoopControlPartialCountMask && !(LC & LoopControlDontUnrollMask)) {
    // If unroll factor is set as '1' - disable loop unrolling
    if (1 == LoopControlParameters[NumParam])
      Metadata.push_back(getMetadataFromName("llvm.loop.unroll.disable"));
    else
      Metadata.push_back(llvm::MDNode::get(
          *Context,
          getMetadataFromNameAndParameter("llvm.loop.unroll.count",
                                          LoopControlParameters[NumParam])));
    ++NumParam;
    assert(NumParam <= LoopControlParameters.size() &&
           "Missing loop control parameter!");
  }

If NumParam == 4 and LoopControlParameters.size() == 4 then assert on line 657 will be passed. But then we may read from memory which is out of LoopControlParameters bound.

AGindinson replied:
The block where this memory is read is not entered unless the correspondent field of the loop control mask is set. This comment does raise an important point, however: I may want to add an assertion somewhere after line 695 to guard against undefined behavior on an invalid SPIR-V.

AGindinson commented:
Applied most of the review comments. Explained non-trivial resolutions in the comments.

AGindinson commented:
TODO:

- [ ] Squash commits and amend several stumps in the main commit message

AlexeySotkin commented on lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -685,49 +685,148 @@ void SPIRVToLLVM::setLLVMLoopMetadata(const LoopInstType *LM, Instruction *BI) {
+          SPIRVId ArraySPIRVId = LoopControlParameters[++NumParam];
+          Value *ArrayVar = ValueMap[M->getValue(ArraySPIRVId)];
+          unsigned Safelen = LoopControlParameters[++NumParam];
+          ArraySflnMap.emplace(std::make_pair(ArrayVar, Safelen));
```
```suggestion
          ArraySflnMap.emplace(ArrayVar, Safelen);
```

AGindinson replied:
Done

AlexeySotkin commented on lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -685,49 +685,148 @@ void SPIRVToLLVM::setLLVMLoopMetadata(const LoopInstType *LM, Instruction *BI) {
+            if (ArraySflnIt != ArraySflnMap.end()) {
+              auto ArrayGEPIt = ArrayGEPMap.find(AccessedArray);
+              if (ArrayGEPIt == ArrayGEPMap.end())
+                ArrayGEPMap.emplace(std::make_pair(
```
```suggestion
                ArrayGEPMap.emplace(ArraySflnIt->first, std::vector<GetElementPtrInst *>{GEP});
```
or maybe
```suggestion
                ArrayGEPMap.emplace(ArraySflnIt->first, {GEP});
```

AGindinson replied:
Reworked to use `operator[]`

AlexeySotkin commented on lib/SPIRV/libSPIRV/spirv.hpp:
```diff
@@ -500,6 +500,7 @@ enum LoopControlMask {
 enum ExtendedControls {
     InitiationIntervalINTEL = 5889,
     MaxConcurrencyINTEL = 5890,
+    DependencyArrayINTEL = 5891,
```
According to the latest version of the spec there is no extended controls anymore
https://github.com/KhronosGroup/SPIRV-Registry/commit/ce3e08be38bdec76a60f4e4a3bef36a1fd2ca51c#diff-2d7da671b991b8c0fd2cde6707831bab

AGindinson replied:
@AlexeySotkin How would you rather I resolved this? Should I include a spec-aligning commit into this PR, or would leaving a `TODO` and tackling it with a new PR suffice?

AlexeySotkin replied:
I'd prefer to align with the spec in scope of a separate PR.

AGindinson replied:
OK, I've added a TODO for this.

AGindinson commented:
Dropped some of the obsolete changes (e.g. moving `hasLoopMetadata()` out of SPIRVInternal.h for no reason whatsoever).
Minimized the code by using `operator[]`.
Squashed the commits. Haven't split into separate commits yet.

AGindinson commented:
Split into two separate commits.

AGindinson commented on lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -685,49 +685,137 @@ void SPIRVToLLVM::setLLVMLoopMetadata(const LoopInstType *LM, Instruction *BI) {
+
+        // Create the index group metadata nodes unique to each array
+        // variables, mark each GEP accessing a particular array variable
+        // a corresponding index group
```
```suggestion
        // into a corresponding index group
```

AlexeySotkin commented on lib/SPIRV/SPIRVWriter.cpp:
```diff
@@ -705,6 +705,156 @@ SPIRV::SPIRVInstruction *LLVMToSPIRV::transUnaryInst(UnaryInstruction *U,
+  unsigned getSafeLen() { return SafeLen; }
+  const std::vector<SPIRVId> &getArrayVariables() { return ArrayVariablesVec; }
+
+  static std::map<MDNode *, SPIRVId> IndexGroupArrayMap;
```
Can we avoid non-const static members?

AGindinson replied:
With pleasure!

AlexeySotkin commented on lib/SPIRV/SPIRVWriter.cpp:
```diff
@@ -705,6 +705,156 @@ SPIRV::SPIRVInstruction *LLVMToSPIRV::transUnaryInst(UnaryInstruction *U,
+        // mask update should happen later, in the main scope of the function
+        unsigned SafeLen = IVDep.getSafeLen();
+        for (auto &ArrayId : IVDep.getArrayVariables())
+          DependencyArrayParameters.push_back(std::make_pair(ArrayId, SafeLen));
```
```suggestion
          DependencyArrayParameters.emplace_back(ArrayId, SafeLen);
```

AGindinson commented on lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -685,49 +685,137 @@ void SPIRVToLLVM::setLLVMLoopMetadata(const LoopInstType *LM, Instruction *BI) {
+        std::map<unsigned, std::vector<MDNode *>> SafelenIdxGroupMap;
+        for (auto &ArrayGEPIt : ArrayGEPMap) {
+          // Emit a distinct index group that will be referenced from
+          // the llvm.loop.parallel_access_indices metadata
```
```suggestion
          // llvm.loop.parallel_access_indices metadata
```

AGindinson commented on lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -685,49 +685,137 @@ void SPIRVToLLVM::setLLVMLoopMetadata(const LoopInstType *LM, Instruction *BI) {
+          }
+        }
+
+        // Create the index group metadata nodes unique to each array
```
```suggestion
        // Create index group metadata nodes specific - one per each array
```

AGindinson commented on lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -685,49 +685,137 @@ void SPIRVToLLVM::setLLVMLoopMetadata(const LoopInstType *LM, Instruction *BI) {
+        }
+
+        // Create the index group metadata nodes unique to each array
+        // variables, mark each GEP accessing a particular array variable
```
```suggestion
        // variables. Mark each GEP accessing a particular array variable
```

AGindinson commented on lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -685,49 +685,137 @@ void SPIRVToLLVM::setLLVMLoopMetadata(const LoopInstType *LM, Instruction *BI) {
+            }
+
+            // If we're dealing with an embedded loop, it may be the case
+            // that the GEP instructions for some of the arrays were already
```
```suggestion
            // that GEP instructions for some of the arrays were already
```

AGindinson commented on lib/SPIRV/SPIRVWriter.cpp:
```diff
@@ -705,6 +705,156 @@ SPIRV::SPIRVInstruction *LLVMToSPIRV::transUnaryInst(UnaryInstruction *U,
+    // casts to an MDNode* rather than an int, 0 will be stored
+    SafeLen = SafeLenExpression ? SafeLenExpression->getZExtValue() : 0;
+
+    // Count the MDNode operands that refer to the index groups:
```
```suggestion
    // Count MDNode operands that refer to index groups:
```

AGindinson commented on lib/SPIRV/SPIRVWriter.cpp:
```diff
@@ -705,6 +705,156 @@ SPIRV::SPIRVInstruction *LLVMToSPIRV::transUnaryInst(UnaryInstruction *U,
+    SafeLen = SafeLenExpression ? SafeLenExpression->getZExtValue() : 0;
+
+    // Count the MDNode operands that refer to the index groups:
+    // - operand [0] is the string literal and should be ignored;
```
```suggestion
    // - operand [0] is a string literal and should be ignored;
```

AGindinson commented on lib/SPIRV/SPIRVWriter.cpp:
```diff
@@ -705,6 +705,156 @@ SPIRV::SPIRVInstruction *LLVMToSPIRV::transUnaryInst(UnaryInstruction *U,
+        // Intel FPGA IVDep loop attribute
+        LLVMParallelAccessIndices IVDep(Node, IndexGroupArrayMap);
+        IVDep.initialize();
+        // Store the IVDep-specific parameters into an intermediate
```
```suggestion
        // Store IVDep-specific parameters into an intermediate
```

AGindinson commented on lib/SPIRV/SPIRVWriter.cpp:
```diff
@@ -705,6 +705,156 @@ SPIRV::SPIRVInstruction *LLVMToSPIRV::transUnaryInst(UnaryInstruction *U,
+        // Store the IVDep-specific parameters into an intermediate
+        // container to address the case when there're multiple
+        // IVDep metadata nodes and this condition gets entered multiple
+        // times. Transfer to the main parameters vector and the loop control
```
```suggestion
        // times. The transfer of these to the main parameters vector & the update
```

AGindinson commented on lib/SPIRV/SPIRVWriter.cpp:
```diff
@@ -705,6 +705,156 @@ SPIRV::SPIRVInstruction *LLVMToSPIRV::transUnaryInst(UnaryInstruction *U,
+        // container to address the case when there're multiple
+        // IVDep metadata nodes and this condition gets entered multiple
+        // times. Transfer to the main parameters vector and the loop control
+        // mask update should happen later, in the main scope of the function
```
```suggestion
        // of the loop control mask will be done later, in the main scope of the function
```
